tests/integration/panic.rs

```diff
@@ -0,0 +1,196 @@
+                fundamental: false,
+                non_enumerable: false,
+                coinductive: false,
+            },
+            associated_ty_ids: vec![],
+            well_known: None,
+        })
+    }
+
+    fn impl_datum(&self, id: ImplId<ChalkIr>) -> Arc<ImplDatum<ChalkIr>> {
+        assert_eq!(id.0.index, 1);
+
+        let substitution = Ty::new(
+            &ChalkIr,
+            ApplicationTy {
+                name: TypeName::Adt(AdtId(RawId { index: 1 })),
+                substitution: Substitution::empty(&ChalkIr),
+            },
+        );
+
+        let binders = Binders::new(
+            VariableKinds::new(&ChalkIr),
+            ImplDatumBound {
+                trait_ref: TraitRef {
+                    trait_id: TraitId(RawId { index: 0 }),
+                    substitution: Substitution::from1(&ChalkIr, substitution),
+                },
+                where_clauses: vec![],
+            },
+        );
+
+        Arc::new(ImplDatum {
+            polarity: Polarity::Positive,
+            binders,
+            impl_type: ImplType::Local,
+            associated_ty_value_ids: vec![],
+        })
+    }
+
+    fn associated_ty_value(
+        &self,
+        id: AssociatedTyValueId<ChalkIr>,
+    ) -> Arc<AssociatedTyValue<ChalkIr>> {
+        unimplemented!()
+    }
+
+    fn opaque_ty_data(&self, id: OpaqueTyId<ChalkIr>) -> Arc<OpaqueTyDatum<ChalkIr>> {
+        unimplemented!()
+    }
+
+    fn adt_datum(&self, id: AdtId<ChalkIr>) -> Arc<AdtDatum<ChalkIr>> {
+        unimplemented!()
+    }
+
+    fn fn_def_datum(&self, fn_def_id: FnDefId<ChalkIr>) -> Arc<FnDefDatum<ChalkIr>> {
+        unimplemented!()
+    }
+
+    fn impls_for_trait(
+        &self,
+        trait_id: TraitId<ChalkIr>,
+        parameters: &[GenericArg<ChalkIr>],
+    ) -> Vec<ImplId<ChalkIr>> {
+        assert_eq!(trait_id.0.index, 0);
+        vec![ImplId(RawId { index: 1 })]
+    }
+
+    fn local_impls_to_coherence_check(&self, trait_id: TraitId<ChalkIr>) -> Vec<ImplId<ChalkIr>> {
+        unimplemented!()
+    }
+
+    fn impl_provided_for(
+        &self,
+        auto_trait_id: TraitId<ChalkIr>,
+        struct_id: AdtId<ChalkIr>,
+    ) -> bool {
+        unimplemented!()
+    }
+
+    fn well_known_trait_id(&self, well_known_trait: WellKnownTrait) -> Option<TraitId<ChalkIr>> {
+        unimplemented!()
+    }
+
+    fn program_clauses_for_env(
+        &self,
+        environment: &Environment<ChalkIr>,
+    ) -> ProgramClauses<ChalkIr> {
+        ProgramClauses::new(&ChalkIr)
+    }
+
+    fn interner(&self) -> &ChalkIr {
+        &ChalkIr
+    }
+
+    fn is_object_safe(&self, trait_id: TraitId<ChalkIr>) -> bool {
+        unimplemented!()
+    }
+}
+
+#[test]
+fn unwind_safety() {
+    use self::MockDatabase;
+    use chalk_integration::interner::{self, ChalkIr};
+    use chalk_ir::*;
+    use std::panic;
+
+    // lower program
+    /*
+    let mut db = lower_program_with_db! {
+        program {
```

Comment:
Let's just make this program a comment by itself :) It's helpful to follow what we're actually writing.

Reply:
I don't extly understand what do you mean - we should leave this program comment, or remove?

I am opting for leaving it, because I find it very helpful understanding what is the goal and things returned from DB about. I think I would move it the module documentation (as DB functions are actually related to it) and maybe describe goal we are proving here.

Reply:
Just add a comment:
```
// goal we are trying to solve
// ...
```

So basically, just remove the extra bits around the program

Reply:
Yea, I reorganized comments a bit so they are more reasonable I think.